### DISCUSSION TITLE
How can I efficently Track the latest version of other's project while keeping mine personal change?

### DISCUSSION THREAD

**Discussion by chenzhigang00** [community/community#136719] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

I downloaded a project from GitHub to my local machine and made some modifications to a few files. However, the project on GitHub has been updated, and I want to sync to the latest version.
 **The thing that bothers me is that I want to keep my own modifications, which are different from other's** 
(e.p if I use 'pull', won't my added content be overwritten?

**Comment by Deekshit7483** [comment]
Hello,

If you make any commits in someone else's repository, those changes will not be directly pushed to that repository. The owner needs to review the changes before they can be merged. However, if you want to see the versions of the commits, go to the original repository and click on 'Commits'.
![Screenshot from 2024-08-24 19-29-51](https://github.com/user-attachments/assets/1567f4bf-bfc3-4933-b0fb-08733c7ee403)
  **Click on commits to view the different versions.**


![Screenshot from 2024-08-24 19-42-22](https://github.com/user-attachments/assets/f3f91d94-2b01-4988-ae87-a867eca9455d)
**You can see all versions of commits**


To download a specific version of the code, click on the desired commit.
![Screenshot from 2024-08-24 19-42-14](https://github.com/user-attachments/assets/b486c082-da40-450e-9363-4fe64eee1886)

Then, click on 'Browse Files,' and you can download the version you selected.

**Comment by eldritch-cookie** [comment, marked as answer]
When you git pull if you have commits it will error and ask you to specify how it should merge,  it can be either rebase or merge.
Rebase removes your commits and then reapplies them on the updated project, merge creates a new commit that has the content of both branches.
In either case if you change a line and the update also changes the same line into different text you will have a conflict which will need to be manually resolved. 
because of conflicts you should have your changes already committed when calling git pull.

**Reply by chenzhigang00** [reply]
Thanks for your quick and kind reply!
But I still have another question, the case is, I want to use other's repository as a tool to finish my own task, and I personalized this tool a little bit. Now there is some new change in the original repository (the owner's), which i want to use but still keep my modification at the same time.
>**That is, I am not making any commits to their's repository at all, just to update the 'tool' to use.**

So, Am I need to firstly  use `fork`, creating my own copy of the repository, then updating my own version, manually compare the differnce between mine and the their's in the end?
Or is there another good way?

**Reply by eldritch-cookie** [reply]
commits are local, so even if in your copy you have a commit it only exists to others when/if you share it. that said if you vehemently don't want a commit you can A) commit your changes, git pull and then use git reset --soft HEAD^ or B) git stash save --all , git pull and then git stash pop

**Reply by chenzhigang00** [reply]
Cool! I got it, thank you so much!

**Reply by ebndev** [reply]
Hey @eldritch-cookie & @chenzhigang00, thanks for taking the time to participate in GitHub Discussions / taking the time to answer!

**Comment by mihailkin** [comment]
Ух ответ и понятный и нет одновременно )
